storage/src/merkle_tree_db_adapter.cpp

```diff
@@ -0,0 +1,323 @@
+}
+
+Sliver DBKeyManipulator::genInternalDbKey(const InternalNodeKey &key) { return serialize(EKeyType::Internal, key); }
+
+Sliver DBKeyManipulator::genStaleDbKey(const InternalNodeKey &key, BlockId staleSinceVersion) {
+  // Use a serialization type to discriminate between internal and leaf keys.
+  return serialize(EKeyType::Stale, staleSinceVersion, KeySerializationType::Internal, key);
+}
+
+Sliver DBKeyManipulator::genStaleDbKey(const LeafKey &key, BlockId staleSinceVersion) {
+  // Use a serialization type to discriminate between internal and leaf keys.
+  return serialize(EKeyType::Stale, staleSinceVersion, KeySerializationType::Leaf, key);
+}
+
+Sliver DBKeyManipulator::generateMetadataKey(ObjectId objectId) { return serialize(EBFTType::Metadata, objectId); }
+
+Sliver DBKeyManipulator::generateStateTransferKey(ObjectId objectId) { return serialize(EBFTType::ST, objectId); }
+
+Sliver DBKeyManipulator::generateSTPendingPageKey(uint32_t pageId) {
+  return serialize(EBFTType::STPendingPage, pageId);
+}
+
+Sliver DBKeyManipulator::generateSTCheckpointDescriptorKey(uint64_t chkpt) {
+  return serialize(EBFTType::STCheckpointDescriptor, chkpt);
+}
+
+Sliver DBKeyManipulator::generateSTReservedPageStaticKey(uint32_t pageId, uint64_t chkpt) {
+  return serialize(EBFTType::STReservedPageStatic, pageId, chkpt);
+}
+
+Sliver DBKeyManipulator::generateSTReservedPageDynamicKey(uint32_t pageId, uint64_t chkpt) {
+  return serialize(EBFTType::STReserverdPageDynamic, pageId, chkpt);
+}
+
+BlockId DBKeyManipulator::extractBlockIdFromKey(const concordUtils::Key &key) {
+  const auto offset = key.length() - sizeof(BlockId);
+  const auto id = concordUtils::netToHost(*reinterpret_cast<const BlockId *>(key.data() + offset));
+
+  LOG_TRACE(
+      logger(),
+      "Got block ID " << id << " from key " << (HexPrintBuffer{key.data(), key.length()}) << ", offset " << offset);
+  return id;
+}
+
+DBAdapter::DBAdapter(const std::shared_ptr<IDBClient> &db, bool readOnly)
+    : DBAdapterBase{db, readOnly}, smTree_{std::make_shared<Reader>(*this)} {}
+
+Status DBAdapter::getKeyByReadVersion(BlockId version,
+                                      const concordUtils::Sliver &key,
+                                      concordUtils::Sliver &outValue,
+                                      BlockId &actualVersion) const {
+  outValue = Sliver{};
+  actualVersion = 0;
+
+  auto iter = db_->getIteratorGuard();
+  const auto dbKey = DBKeyManipulator::genDataDbKey(key, version);
+  const auto &[foundKey, foundValue] = iter->seekAtLeast(dbKey);
+  if (iter->isEnd()) {
+    const auto &[prevKey, prevValue] = iter->previous();
+    if (!prevKey.empty()) {
+      outValue = prevValue;
+      actualVersion = DBKeyManipulator::extractBlockIdFromKey(prevKey);
+    }
+  } else if (foundKey == dbKey) {
+    outValue = foundValue;
+    actualVersion = DBKeyManipulator::extractBlockIdFromKey(foundKey);
+  }
+
+  return Status::OK();
```

Comment:
I really think we misuse status responses here. I realize that this likely matches up with existing behavior, but it's probably worth fixing in the future to return `Status::NotFound` when a key is not found rather than `Status::OK`. Then we'd have a more explicit check rather than checking if the key is empty.

Reply:
Yes, totally agree.